clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -331,9 +335,38 @@ FuncType FuncType::clone(TypeRange inputs, TypeRange results) const {
   return get(llvm::to_vector(inputs), results[0], isVarArg());
 }
 
-mlir::ParseResult parseFuncTypeArgs(mlir::AsmParser &p,
-                                    llvm::SmallVector<mlir::Type> &params,
-                                    bool &isVarArg) {
+// A special parser is needed for function returning void to handle the missing
+// type.
+static mlir::ParseResult parseFuncTypeReturn(mlir::AsmParser &p,
+                                             mlir::Type &optionalReturnType) {
+  if (succeeded(p.parseOptionalLParen())) {
+    // If we have already a '(', the function has no return type
+    optionalReturnType = {};
+    return mlir::success();
+  }
+  mlir::Type type;
+  if (p.parseType(type))
+    return mlir::failure();
+  if (isa<cir::VoidType>(type))
+    // An explicit !cir.void means also no return type.
+    optionalReturnType = {};
+  else
+    // Otherwise use the actual type.
+    optionalReturnType = type;
+  return p.parseLParen();
+}
+
+// A special pretty-printer for function returning or not a result.
+static void printFuncTypeReturn(mlir::AsmPrinter &p,
+                                mlir::Type optionalReturnType) {
+  if (optionalReturnType)
+    p << optionalReturnType << ' ';
+  p << '(';
+}
+
+static mlir::ParseResult
+parseFuncTypeArgs(mlir::AsmParser &p, llvm::SmallVector<mlir::Type> &params,
+                  bool &isVarArg) {
   isVarArg = false;
   // `(` `)`
   if (succeeded(p.parseOptionalRParen()))
```

Comment:
I see what is happening here now, not sure I like it?  IMO, the parse-args function should be responsible for 'taking' BOTH of the parens, everything else should do a 'look-ahead' type parsing if necessary.  THOUGH I see the latter doesn't seem to exist in `mlir::AsmParser`?  Is this a case of us 'using it wrong'? What do other parsers do here?

Reply:
I am not familiar with `mlir::AsmParser`, but looking through its definition just now I don't see anyway to peek at the next token or do lookahead or do a trial parse without advancing.  I agree that `parseFuncTypeArgs` should handle both the `(` and the `)`.  But I don't see a way to do that.  I don't have experience with other MLIR parsers, so I don't know what they do.  Unless someone more knowledgeable pipes up, I don't see a way to change this.

Reply:
@lanza @bcardosolopes Anyone have any ideas here?  This parsing is REALLY weird, but we still need a way to know that we're 'done'.

The other alternative is to change how these are printed/read (so it is something like: `(returnType)(Arg1, Arg2)` with whatever delimiters make sense), but I'm not sure that is something that can be done sensibly.

Reply:
I think this is the way the parsers are implemented to chain things ahead.
See again my comment with `void` keyword.
At the end it is a trade-off between implementator comfort and code beauty vs the concrete assembly syntax exposed to the end user.

Reply:
You can technically do it the other way around as mlir function types are structured and have `!cir.func<(args) -> ret>` or `!cir.func<(args)>` in case of void. This can be parsed without lookahead using `parseOptionalArrow`.

Reply:
`(args)` can be nicely parsed using `parseCommaSeparatedList(mlir::AsmParser::Delimiter::Paren, parseArg)` which takes care of both `parens`

Reply:
I like the idea of the `!cir.func<(args) -> ret>` format. It feels more consistent with func dialect.

Reply:
I like this idea too.
I have tried to be as close as the original syntax to avoid disrupting too much the tests and the users but there is no reason to do something else.
If there is a consensus from the ClangIR stake-holders, go for it!

Reply:
@bcardosolopes @lanza ^

Reply:
The only question is what the 'void' type looks like?  Could/should we still use the `void` as the return type in the textual representation, or alternatively just omit entirely? So:

`!cir.func<(args) -> void>`
vs: 
`!cir.func<(args)>`

That way we do `parse arguments`, then `if arrow, parse return type, else is void`.

The 'trailing return type' syntax is nice as it is consistent with the C++ language.

Reply:
Got it, thanks for the update!  I was trying to judge opinions here, but the incubator review is sufficient to make sure those opinions are heard.

Reply:
> I like the idea of the !cir.func<(args) -> ret> format

+1